### ISSUE TITLE
Ability to merge state changes instead of fully replacing

### ISSUE THREAD

**Issue by WesCossick** [ReactTraining/history#698]
We're working on the ability to easily store and recover state as a user navigates through a SPA. We started by using this package's `.replace()` function, which, as long as you're replacing the entire state, works fine.

But assuming there are multiple pieces of data that we want to store in the history state for a single location, and assuming the logic for storing each piece of data is isolated from one another, you need to merge the current state with what you want to change, and then pass the merged object to `.replace()`. That way you don't eliminate the other pieces of data.

However, because `.replace()` executes asynchronously, the state isn't fully updated right away. So subsequent calls to `.replace()` that try to merge using the current state end up using stale data (if those calls happen quickly enough).

Here's an example. Let's say you want to keep the following two pieces of data in the history state for a particular page:

```json
{
    "piece1": "original_value1",
    "piece2": "original_value2"
}
```

And let's say you want to update the state in two separate locations in rapid succession. I'll use [deepmerge](https://www.npmjs.com/package/deepmerge) for this example.

```js
// In one place in the code...
history.replace(location.pathname, deepmerge(location.state, {
    piece1: 'new_value1'
});

// In another place in the code...
history.replace(location.pathname, deepmerge(location.state, {
    piece2: 'new_value2'
});
```

At the end of this code, the state would contain:

```json
{
    "piece1": "original_value1",
    "piece2": "new_value2"
}
```

We're currently solving this issue by interfacing with the browser's history API directly. It appears to execute synchronously, so there's no timing issue. But since we're using React Router, it would be less hacky to interact directly with this package's API.

The easiest way to solve this, I believe, would be to introduce an option that asked for the state to be merged instead of fully replaced. If the merging happened at the right point in this package's codebase, it should eliminate the problem entirely. You could use [deepmerge](https://www.npmjs.com/package/deepmerge), even. It has no dependencies, so it wouldn't bloat the codebase.

**Comment by StringEpsilon** [comment]
I'm not entirely convinced of the usecase, but I have a more versatile approach to solving it:

```replace``` could take a callback as an additional argument that is responsible of assigning the new state value based on the old and the new state.

```js
history.replace("/foo", { a: "foo" }); 
// location state = { a:  "foo" }
history.replace("/foo", { b: "bar" }, (old, new) => ({ ...old, ...new }) );
// location state = { a:  "foo", b: "bar" } 
``` 

That way, people can make write their own logic and history doesn't have a ~10% bundle size increase:

|    | min | minzip 
|---|---|---
|history| 10.1kB | 3.6 kB
|deepmerge| 1.3kB | 620 B

**Comment by WesCossick** [comment]
@StringEpsilon That could definitely work too, and I like its added flexibility.

The real key is that the execution has to be synchronous between the point where the original state is retrieved/merged, and the point where the new state is actually stored in the browser history. Right now, merging happens and then a (presumably) asynchronous callback handles storing it in the browser history. Between those two points, state changes from other sources in the code can occur, leading to stale data overwriting recent changes.

**Comment by shavyg2** [comment]
If history was implement this they could use immer to keep it the state from changing as well

**Comment by mjackson** [comment]
@WesCossick If I understand correctly it seems like the whole "transition manager" is screwing things up for you. Is that right?

**Comment by WesCossick** [comment]
@mjackson I'm not completely certain what you mean by "transition manager," but as mentioned in [your comment](https://github.com/ReactTraining/history/pull/699#issuecomment-518764782), I believe if the logic executed synchronously instead of asynchronously, it could solve inherently solve the issue.

**Comment by WesCossick** [comment]
> how are we supposed to know you want to merge the two?

@mjackson, see:

> The easiest way to solve this, I believe, would be to introduce an option that asked for the state to be merged instead of fully replaced. If the merging happened at the right point in this package's codebase, it should eliminate the problem entirely. You could use [deepmerge](https://www.npmjs.com/package/deepmerge), even. It has no dependencies, so it wouldn't bloat the codebase.

**Comment by WesCossick** [comment]
> If the keys collide, should we overwrite the same keys from the first call?

Yes, that's exactly what a "merge" does. It overwrites any existing keys with new values.
